### PR TITLE
Add MAX31865 sensor support, fix MAX31855 sensor

### PR DESCRIPTION
**Pull request by DAVe3283**
### MAX31865

Added support for the MAX31865 RTD-to-Digital Converter to measure PT100 and similar RTDs. Verified with an Adafruit unit (product ID: 3328) and a PT100 probe.

### MAX31855

This was setup for incorrect SPI clock polarity and phase, and would return bad data due to a race condition measuring on the wrong edge (verified with Saleae Logic scope). Selecting the correct configuration fixes that problem.

Re-wrote the decode off the datasheet to handle error states better (sends NaN as an update on failure to read temperature, which shows the value as Unknown in Home Assistant).

Added the *optional* ability to monitor the internal high-precision temperature sensor, which can be nice in some applications.

**Related issue (if applicable):** N/A

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#399

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Review by sylviamic — APPROVED** [review]
This looks nice, and appears to work with my own MAX31865 sensor. Great work! Looking forward to seeing this merged.

**Comment by DAVe3283** [comment]
I am using this code successfully to control a chilled water bath. The ESP32 and MAX31865 (and MAX31855 before it) are located in the same housing as the chiller compressor, which is a very electrically noisy environment. These latest changes to fault detection have the sensor operating very smoothly, handling the occasional overvoltage fault when the compressor kicks on.

I am very confident in this code for both the MAX31855 and MAX31865, as I have now run both for several days with no software issues. I can report back again in another few days if desired, but I feel this code is ready to merge.

File: esphome/components/max31855/max31855.cpp [hunk 1]
```diff
@@ -22,63 +23,84 @@ void MAX31855Sensor::setup() {
...
 }
 float MAX31855Sensor::get_setup_priority() const { return setup_priority::DATA; }
 void MAX31855Sensor::read_data_() {
   this->enable();
   delay(1);
   uint8_t data[4];
   this->read_array(data, 4);
+  this->disable();
+
+  const uint32_t mem(data[0] << 24 | data[1] << 16 | data[2] << 8 | data[3] << 0);
```
Comment (OttoWinter) [inline esphome/components/max31855/max31855.cpp]:
```suggestion
  const uint32_t mem = data[0] << 24 | data[1] << 16 | data[2] << 8 | data[3] << 0;
```

I'm not sure why you used the `()` constructor syntax for declaring variables - but IMO it's much harder to read. I strongly recommend using the equals syntax.

(Applies in other places too).

File: esphome/components/max31855/max31855.cpp [hunk 2]
```diff
@@ -22,63 +23,84 @@ void MAX31855Sensor::setup() {
...
 
-  // val is 14 bits of signed temperature data followed by 2 bits of status flags
-  int16_t val = data[1] | data[0] << 8;
+  // Verify we got data
+  if (mem && mem != 0xFFFFFFFF) {
```
Comment (OttoWinter) [inline esphome/components/max31855/max31855.cpp]:
```suggestion
  if (mem != 0 && mem != 0xFFFFFFFF) {
```

Style suggestion.

File: esphome/components/max31855/sensor.py [hunk 3]
```diff
@@ -1,14 +1,16 @@
...
-from esphome.const import CONF_ID, ICON_THERMOMETER, UNIT_CELSIUS
+from esphome.const import CONF_ID, CONF_REFERENCE_TEMPERATURE, ICON_THERMOMETER, UNIT_CELSIUS
 
 max31855_ns = cg.esphome_ns.namespace('max31855')
 MAX31855Sensor = max31855_ns.class_('MAX31855Sensor', sensor.Sensor, cg.PollingComponent,
                                     spi.SPIDevice)
 
 CONFIG_SCHEMA = sensor.sensor_schema(UNIT_CELSIUS, ICON_THERMOMETER, 1).extend({
     cv.GenerateID(): cv.declare_id(MAX31855Sensor),
+    cv.Optional(CONF_REFERENCE_TEMPERATURE):
```
Comment (OttoWinter) [inline esphome/components/max31855/sensor.py]: Usually, I would not allow this syntax (sensor schema within sensor schema) - should be base schema with two sensor schemas in it (see also dht). But it's ok here for avoiding a breaking change 👍

File: esphome/components/max31855/max31855.cpp (hunk 1 above)
Comment (DAVe3283) [reply esphome/components/max31855/max31855.cpp]: Just habit from other projects. I will happily update to match the project style later tonight.

**Comment by DAVe3283** [comment]
The MAX31865 sensor has been working great all week, enabling a climate component to run the chiller.
Soda Fountain
The output of this sensor is so good I've got the water bath very close to freezing with no issues, even in the noisy electrical environment it has to deal with.
Water Temperature

**Review by OttoWinter — APPROVED** [review]

**Pull request merged**